### PR TITLE
[flutter_tools] scaffolding for Windows UWP template

### PR DESCRIPTION
**Pull request by jonahwilliams**
Adds all the scaffolding that will allow @clarkezone to easily drop in their windows UWP project templates

Work towards https://github.com/flutter/flutter/issues/14967

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/create.dart [hunk 1]
```diff
@@ -22,15 +22,18 @@ import '../reporting/reporting.dart';
 import '../runner/flutter_command.dart';
 import 'create_base.dart';
 
+const String kPlatformHelp =
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create.dart:25]: minor refactoring

File: packages/flutter_tools/lib/src/commands/create_base.dart [hunk 2]
```diff
@@ -130,17 +131,28 @@ abstract class CreateBase extends FlutterCommand {
   @protected
   void addPlatformsOptions({String customHelp}) {
     argParser.addMultiOption('platforms',
-        help: customHelp ?? _kDefaultPlatformArgumentHelp,
-        defaultsTo: _kAvailablePlatforms,
-        allowed: _kAvailablePlatforms);
+      help: customHelp ?? _kDefaultPlatformArgumentHelp,
+      defaultsTo: <String>[
+        ..._kAvailablePlatforms,
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create_base.dart:136]: make sure winuwp is conditionally added so it doesn't show up as an option unless you've already opted in

File: packages/flutter_tools/lib/src/commands/create_base.dart [hunk 3]
```diff
@@ -156,37 +168,8 @@ abstract class CreateBase extends FlutterCommand {
   }
 
   /// Gets the flutter root directory.
-  ///
-  /// Throw with exit code 2 if the flutter sdk installed is invalid.
   @protected
-  String get flutterRoot {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create_base.dart:162]: This is completely unnecessary and I couldn't figure out what it is doing :)

File: packages/flutter_tools/lib/src/commands/create_base.dart [hunk 4]
```diff
@@ -616,32 +604,10 @@ const Set<String> _keywords = <String>{
 };
 
 const Set<String> _packageDependencies = <String>{
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create_base.dart:606]: Many of these are no longer in the basic flutter dependency graph, so I removed them

File: packages/flutter_tools/templates/template_manifest.json [hunk 5]
```diff
@@ -137,6 +137,7 @@
         "templates/app/windows.tmpl/runner/utils.h",
         "templates/app/windows.tmpl/runner/win32_window.cpp",
         "templates/app/windows.tmpl/runner/win32_window.h",
+        "templates/app/winuwp.tmpl/data/version",
```
Comment (jonahwilliams) [inline packages/flutter_tools/templates/template_manifest.json:140]: Rest of the template files would need to be added here too

**Comment by jonahwilliams** [comment]
hmm, need to figure out what I misconfigured

**Comment by jonahwilliams** [comment]
Failures are due to CI issues with gradle

File: packages/flutter_tools/templates/app/winuwp.tmpl/project_version [hunk 6]
```diff
@@ -0,0 +1 @@
+0
```
Comment (jonahwilliams) [inline packages/flutter_tools/templates/app/winuwp.tmpl/project_version:1]: This is the folder where the rest of the template files need to go. There is also a manifest in the templates/ where they need to be listed.

**Comment by clarkezone** [comment]
I will take a look this evening.  Made good progress on the cmake port of the UWP template over the weekend should have an initial version in a few days

**Review by clarkezone — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/commands/create_base.dart (hunk 3 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/create_base.dart:162]: A wild guess, but Is it needed for --local-engine? Are there checks elsewhere that `throwToolExit` when `Cache.flutterRoot` is null? It looks like these are very conservative sanity checks, were they added in response to a github issue?

File: packages/flutter_tools/lib/src/commands/create.dart [hunk 7]
```diff
@@ -391,6 +394,7 @@ Your $application code is in $relativeAppMain.
       templateContext['linux'] = false;
       templateContext['macos'] = false;
       templateContext['windows'] = false;
+      templateContext['winuwp'] = false;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/create.dart:397]: Maybe a list of platforms would help clean this up, like:
```dart
for (final String platform in kPlatforms) {
  templateContext[platform] = false;
}
```

File: packages/flutter_tools/lib/src/commands/create.dart [hunk 8]
```diff
@@ -469,6 +473,8 @@ Your $application code is in $relativeAppMain.
         'windows',
       if (templateContext['macos'] == true)
         'macos',
+      if (templateContext['winuwp'] == true)
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/create.dart:476]: Ditto:
```dart
  for (final String platform in kPlatforms)
    if (templateContext[platform] == true)
      platform,
```

File: packages/flutter_tools/lib/src/commands/create_base.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/create_base.dart:162]: localEngine doesn't modify FlutterRoot though 🤔

File: packages/flutter_tools/lib/src/commands/create.dart (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/create.dart]: done

File: packages/flutter_tools/lib/src/commands/create.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/create.dart]: done

File: packages/flutter_tools/lib/src/commands/create_base.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/create_base.dart:162]: I traced it back as far as https://github.com/flutter/flutter/commit/cccd917ac25b043b84449096ccbe51ab0492a85c

My guess is we used to be less careful about setting the flutter root to something valid, but the logic is much more solid (and tested !) now

**Review by zanderso — APPROVED** [review]

**Pull request merged**